// @matcha2d/types/src/config.ts
export interface WorldConfig {
  gravity: { x: number; y: number }
  fixedTimestep: number
  /** Outer solver substeps per `engine.step()` (integration between passes). */
  velocityIterations: number
  /** PGS (Gauss–Seidel) iterations per substep; higher improves contact/friction convergence. */
  internalVelocityIterations: number
  /** @deprecated Ignored by the WASM solver (stabilization pass removed). Kept for saved configs / API stability. */
  positionIterations: number
  /** Position correction bias factor (0.0–1.0). 0.05 is conservative; 0.2 is Box2D's default for meter-scale physics. */
  baumgarteFactor: number
  /**
   * Maximum velocity magnitude for spring penetration bias (`f32::INFINITY` / `Number.POSITIVE_INFINITY` = uncapped).
   */
  maxCorrectiveVelocity: number
  /** Penetration slop: allowed penetration depth before correction kicks in. */
  penetrationSlop: number
  /** @deprecated Ignored by the WASM solver (per-substep impulse reset). Kept for saved configs / API stability. */
  warmStarting: boolean
  /** Enable block solver for 2-contact manifolds (reduces jitter). */
  blockSolver: boolean
  /** Default friction coefficient when not specified per-body. */
  defaultFriction: number
  /** Default restitution (bounciness) coefficient when not specified per-body. */
  defaultRestitution: number
  /** Restitution slop: velocity threshold below which restitution is ignored. */
  restitutionSlop: number
  /** Default linear damping coefficient (energy dissipation). 0 = no damping, 1 = heavy damping. */
  defaultLinearDamping: number
  /** Default angular damping coefficient. */
  defaultAngularDamping: number
  /** Enable swept AABB broadphase to catch fast-moving objects (CCD-lite). */
  sweptBroadphase: boolean
  /** Linear velocity threshold (m/s) below which a body is a candidate for sleeping. */
  sleepVelocityThreshold: number
  /** Angular velocity threshold (rad/s) below which a body is a candidate for sleeping. */
  sleepAngularVelocityThreshold: number
  /** Time (seconds) a body must be below velocity thresholds before it sleeps. */
  sleepTimeThreshold: number
  /**
   * Optional post-PGS clamp on |linear velocity| for bodies in contact (`Infinity` disables).
   * Prefer raising substeps / `pgsRelaxation` tuning over low caps.
   */
  maxSolverLinearVelocity: number
  /** Optional post-PGS clamp on |angular velocity| (`Infinity` disables). */
  maxSolverAngularVelocity: number
  /** PGS impulse under-relaxation in `(0, 1]` — principal stability knob instead of velocity clamps. */
  pgsRelaxation: number
  /**
   * Optional cap on |linear velocity| before substep integration (`Infinity` = none).
   * Does not replace CCD / TOI for tunneling.
   */
  maxIntegrateLinearVelocity: number
  /** Optional cap on |angular velocity| before substep integration (`Infinity` = none). */
  maxIntegrateAngularVelocity: number
  /** Optional axis-aligned bounds (meters); dynamic bodies are clamped after each step. Pass null to disable. */
  worldBounds: { minX: number; minY: number; maxX: number; maxY: number } | null
}

export const DEFAULT_WORLD_CONFIG: Readonly<WorldConfig> = {
  gravity: { x: 0, y: -9.81 },
  fixedTimestep: 1 / 60,
  velocityIterations: 12,
  internalVelocityIterations: 8,
  positionIterations: 6,
  baumgarteFactor: 0.2,
  maxCorrectiveVelocity: Number.POSITIVE_INFINITY,
  penetrationSlop: 0.02,
  warmStarting: true,
  blockSolver: true,
  defaultFriction: 0.3,
  defaultRestitution: 0.0,
  restitutionSlop: 1.0,
  defaultLinearDamping: 0.01,
  defaultAngularDamping: 0.01,
  sweptBroadphase: true,
  sleepVelocityThreshold: 0.25,
  sleepAngularVelocityThreshold: 0.5,
  sleepTimeThreshold: 1.0,
  maxSolverLinearVelocity: Number.POSITIVE_INFINITY,
  maxSolverAngularVelocity: Number.POSITIVE_INFINITY,
  pgsRelaxation: 0.88,
  maxIntegrateLinearVelocity: Number.POSITIVE_INFINITY,
  maxIntegrateAngularVelocity: Number.POSITIVE_INFINITY,
  worldBounds: null,
}
