// @matcha2d/render/src/canvas2d.ts
import type { MatchaBuffers, NarrowphaseColliderOutline, ShapeType } from '@matcha2d/types'
import type { IRenderer } from './renderer.js'

const BodyFlags = {
  ACTIVE: 0x01,
  STATIC: 0x02,
  SLEEPING: 0x04,
  SENSOR: 0x08,
} as const

/**
 * Canvas2D debug renderer.
 * Draws oriented bodies (circles, boxes, polygons) with proper rotation.
 * Following the article's guidance: shapes are drawn in world space by
 * applying rotation matrices to transform from model space to world space.
 */
export class Canvas2DRenderer implements IRenderer {
  private _scale = 50 // pixels per meter
  private _offsetX = 0
  private _offsetY = 0

  constructor(
    private readonly _ctx: CanvasRenderingContext2D,
    options?: { scale?: number; offsetX?: number; offsetY?: number }
  ) {
    if (options?.scale) this._scale = options.scale
    if (options?.offsetX) this._offsetX = options.offsetX
    if (options?.offsetY) this._offsetY = options.offsetY
  }

  /** Set the camera scale (pixels per meter). */
  setScale(scale: number): void {
    this._scale = scale
  }

  /** Set the camera offset in pixels. */
  setOffset(x: number, y: number): void {
    this._offsetX = x
    this._offsetY = y
  }

  begin(): void {
    this._ctx.clearRect(0, 0, this._ctx.canvas.width, this._ctx.canvas.height)
    this._ctx.save()
    // Flip Y axis so positive Y is up (physics convention)
    this._ctx.translate(this._offsetX, this._ctx.canvas.height + this._offsetY)
    this._ctx.scale(this._scale, -this._scale)
  }

  drawBodies(buffers: MatchaBuffers, count: number, _alpha: number): void {
    const {
      positionX, positionY,
      angle,
      shapeType, shapeRadius,
      halfExtentX, halfExtentY,
      shapeVertexCount, shapeVerticesX, shapeVerticesY,
      flags,
    } = buffers

    for (let i = 0; i < count; i++) {
      // Skip inactive bodies
      if (!(flags[i] & BodyFlags.ACTIVE)) continue

      const px = positionX[i]
      const py = positionY[i]
      const rot = angle[i]
      const shape = shapeType[i] as ShapeType

      // Choose color based on body state
      const isStatic = (flags[i] & BodyFlags.STATIC) !== 0
      const isSleeping = (flags[i] & BodyFlags.SLEEPING) !== 0
      const isSensor = (flags[i] & BodyFlags.SENSOR) !== 0

      this._ctx.save()
      this._ctx.translate(px, py)
      this._ctx.rotate(rot)

      if (isSensor) {
        this._ctx.globalAlpha = 0.4
      }

      switch (shape) {
        case 0: // Box
          this._drawBox(halfExtentX[i], halfExtentY[i], isStatic, isSleeping)
          break
        case 1: // Circle
          this._drawCircle(shapeRadius[i], isStatic, isSleeping)
          break
        case 2: // Polygon
          this._drawPolygon(i, shapeVertexCount[i], shapeVerticesX, shapeVerticesY, isStatic, isSleeping)
          break
      }

      this._ctx.restore()
    }
  }

  end(): void {
    this._ctx.restore()
  }

  /**
   * Draw narrowphase-accurate collider outlines in **world meters** (Y-up).
   * Call between `begin()` and `end()` so the same transform as `drawBodies` applies.
   */
  drawNarrowphaseColliderOutlinesMeters(
    outlines: NarrowphaseColliderOutline[],
    options?: { lineWidth?: number; strokeSolid?: string; strokeSensor?: string },
  ): void {
    const ctx = this._ctx
    const lw = (options?.lineWidth ?? 2) / this._scale
    const strokeSolid = options?.strokeSolid ?? 'rgba(0, 255, 140, 0.9)'
    const strokeSensor = options?.strokeSensor ?? 'rgba(255, 210, 60, 0.85)'
    for (const o of outlines) {
      if (o.vertices.length === 0) continue
      ctx.strokeStyle = o.sensor ? strokeSensor : strokeSolid
      ctx.lineWidth = lw
      ctx.beginPath()
      for (let v = 0; v < o.vertices.length; v++) {
        const p = o.vertices[v]
        if (v === 0) ctx.moveTo(p.x, p.y)
        else ctx.lineTo(p.x, p.y)
      }
      ctx.closePath()
      ctx.stroke()
    }
  }

  destroy(): void {
    // Canvas2D doesn't need cleanup
  }

  private _drawBox(hx: number, hy: number, isStatic: boolean, isSleeping: boolean): void {
    const ctx = this._ctx

    ctx.strokeStyle = isStatic ? '#444' : isSleeping ? '#888' : '#2196F3'
    ctx.lineWidth = 2 / this._scale
    ctx.fillStyle = isStatic ? '#666' : isSleeping ? '#aaa' : 'rgba(33, 150, 243, 0.3)'

    ctx.beginPath()
    ctx.rect(-hx, -hy, hx * 2, hy * 2)
    ctx.fill()
    ctx.stroke()

    // Draw orientation indicator (line from center to right edge)
    ctx.strokeStyle = isStatic ? '#888' : '#FF5722'
    ctx.lineWidth = 1.5 / this._scale
    ctx.beginPath()
    ctx.moveTo(0, 0)
    ctx.lineTo(hx * 0.8, 0)
    ctx.stroke()
  }

  private _drawCircle(radius: number, isStatic: boolean, isSleeping: boolean): void {
    const ctx = this._ctx

    ctx.strokeStyle = isStatic ? '#444' : isSleeping ? '#888' : '#4CAF50'
    ctx.lineWidth = 2 / this._scale
    ctx.fillStyle = isStatic ? '#666' : isSleeping ? '#aaa' : 'rgba(76, 175, 80, 0.3)'

    ctx.beginPath()
    ctx.arc(0, 0, radius, 0, Math.PI * 2)
    ctx.fill()
    ctx.stroke()

    // Draw orientation indicator (line from center to edge)
    ctx.strokeStyle = isStatic ? '#888' : '#FF5722'
    ctx.lineWidth = 1.5 / this._scale
    ctx.beginPath()
    ctx.moveTo(0, 0)
    ctx.lineTo(radius * 0.8, 0)
    ctx.stroke()
  }

  private _drawPolygon(
    bodyIdx: number,
    vertexCount: number,
    verticesX: Float32Array,
    verticesY: Float32Array,
    isStatic: boolean,
    isSleeping: boolean
  ): void {
    if (vertexCount === 0) return

    const ctx = this._ctx
    const stride = 16 // MAX_VERTICES_PER_SHAPE from types
    const base = bodyIdx * stride

    ctx.strokeStyle = isStatic ? '#444' : isSleeping ? '#888' : '#9C27B0'
    ctx.lineWidth = 2 / this._scale
    ctx.fillStyle = isStatic ? '#666' : isSleeping ? '#aaa' : 'rgba(156, 39, 176, 0.3)'

    ctx.beginPath()
    for (let v = 0; v < vertexCount; v++) {
      const vx = verticesX[base + v]
      const vy = verticesY[base + v]
      if (v === 0) {
        ctx.moveTo(vx, vy)
      } else {
        ctx.lineTo(vx, vy)
      }
    }
    ctx.closePath()
    ctx.fill()
    ctx.stroke()

    // Draw orientation indicator (line from center to first vertex)
    const firstVx = verticesX[base]
    const firstVy = verticesY[base]
    const len = Math.sqrt(firstVx * firstVx + firstVy * firstVy)
    if (len > 0.001) {
      ctx.strokeStyle = isStatic ? '#888' : '#FF5722'
      ctx.lineWidth = 1.5 / this._scale
      ctx.beginPath()
      ctx.moveTo(0, 0)
      ctx.lineTo(firstVx * 0.8, firstVy * 0.8)
      ctx.stroke()
    }
  }
}
